Line segment intersection

jbeenie opened the pull request.

jeremiebenhamron committed:
add Initializer to Line struct and document it.

jeremiebenhamron committed:
Fix Description of angle variable in Line struct

jeremiebenhamron committed:
add precondtion to LineSegment intializer, start and end points must be different

jeremiebenhamron committed:
add 2 dimensional cross product of 2 vectors

jeremiebenhamron committed:
add areParallel static method to Vector

jeremiebenhamron committed:
make unit vector initalizer a static method

jeremiebenhamron committed:
add unitvector computed var to vector

jeremiebenhamron committed:
implement intersection of 2 line segment algorithm

jeremiebenhamron committed:
add conditional conformance to ApproximatelyEquatable for Arrays with elements that are approximately equatable.

jeremiebenhamron committed:
make Intersection conform to Approximately Equatable

jeremiebenhamron committed:
fix typo

jeremiebenhamron committed:
add are parallel tests

jeremiebenhamron committed:
Add LineSegment intersection Tests

alejandro-isaza commented on PathKit/Sources/Line.swift:
```diff
@@ -9,21 +9,32 @@ import Foundation
     /// - Returns: True if `point` is on `self` within the given accuracy.
     public func contains(_ point: Point, accuracy: Double) -> Bool {
-        let dotProduct = Vector(head: point, tail: .zero) • Vector(angle: angle)
+        let dotProduct = Vector(head: point, tail: .zero) • Vector.unitVector(with: angle)
```
Better!

alejandro-isaza commented on PathKit/Sources/LineSegment.swift:
```diff
@@ -27,13 +27,64 @@ public struct LineSegment {
     ///   - end: The end point of the line segment.
+    /// - Precondition: The `start` point and `end` point of the line segment must be different.
     public init(start: Point, end: Point) {
+        precondition(start != end)
```
Why?

jbeenie replied:
It avoids division by zero in a couple of places.

alejandro-isaza commented on PathKit/Sources/LineSegment.swift:
```diff
@@ -27,13 +27,64 @@ public struct LineSegment {
+    /// Source: https://stackoverflow.com/questions/563198/how-do-you-detect-where-two-line-segments-intersect
+    public static func intersection(_ lhs: LineSegment, _ rhs: LineSegment, accuracy: Double) -> Intersection {
+        let p = lhs.start
+        let q = rhs.start
```
do we really need to add these `let`s?

jbeenie replied:
will remove these

alejandro-isaza commented on PathKit/Sources/LineSegment.swift:
```diff
@@ -27,13 +27,64 @@ public struct LineSegment {
+    public static func intersection(_ lhs: LineSegment, _ rhs: LineSegment, accuracy: Double) -> Intersection {
+        let p = lhs.start
+        let q = rhs.start
+        let qp = Vector(head: q, tail: p)
```
Let's come up with a better name for this. Something like `headsVector`

jbeenie replied:
ok will do

alejandro-isaza commented on PathKit/Sources/LineSegment.swift:
```diff
@@ -27,13 +27,64 @@ public struct LineSegment {
+        let q = rhs.start
+        let qp = Vector(head: q, tail: p)
+        let r = Vector(head: lhs.end, tail: lhs.start)
+        let s = Vector(head: rhs.end, tail: rhs.start)
```
`lhsVector` and `rhsVector`?

alejandro-isaza replied:
Also why are they backwards?

jbeenie replied:
They're not backwards, the vector points in the direction  tail -> head. So the tail is like the start of the vector and the head is like the end.

alejandro-isaza replied:
I get confused by this every time...

jbeenie replied:
:)

alejandro-isaza commented on PathKit/Sources/LineSegment.swift:
```diff
@@ -27,13 +27,64 @@ public struct LineSegment {
+            let t = qp ✕ s / rXs
+            let u = qp ✕ r / rXs
+
+            if zeroToOne.contains(t) && zeroToOne.contains(u) {
```
I think the accuracy should be taken into account here?

jbeenie replied:
Good catch. Will fix

jeremiebenhamron committed:
Move logic into line segment extension

jeremiebenhamron committed:
add vector computed variable to line segment

jeremiebenhamron committed:
update variable names in linesegment  intersection algorithm

jeremiebenhamron committed:
update to swift 5

jeremiebenhamron committed:
update redundant access modifiers

jeremiebenhamron committed:
make linesegment equatable and approximately equatable.

jeremiebenhamron committed:
create Overlap struct

jeremiebenhamron committed:
use overlap in infinite case of intersection.

jeremiebenhamron committed:
Move extensions into seperate files

jeremiebenhamron committed:
Compute overlap of intersecting line segments.

jbeenie commented:
I added the logic computing the overlap of intersecting LineSegments. Need to add units tests now to test that case and tests a couple other functions I added.

jeremiebenhamron committed:
Test that intersection of line returns appropriate overlap.

jeremiebenhamron committed:
set language version for project.

jeremiebenhamron committed:
update travis file to test code using xcode 10.2
